libhb: Fixes and enhancements for qsv encoding with HEVC

DOram424 opened the pull request:
The motivation for these patches was that the sample_encode executable that is part of the  mfx libraries was producing smaller files with similar quality and these patches bring handbrake into line with minimal changes.

There's a tonne of other stuff I can add in here such as presets / support for CodingOption3, proper parameter rejection and support detection, etc. etc. but I've kept it minimal and done my best to avoid altering code structure and the GUI. I'm sure these limited patches probably have more than enough horrific code blockers and shoddy C coding to be getting on with and we can come back to the rest if it's useful / desired (to be discussed on the forums). Given the complexity in configuration especially cross-parameter requirements for the B tree that includes parameters that can't even be configured I've set a default configuration so that users only need configure the qp value, codec and bit rate method.

I've mainly used linux, but have also tested hevc under windows using command line only. If --enable-qsv-decoding is used then --qsv-async-depth=1 is a requirement for windows. I think enable-qsv-decoding was the default .. but as I've only just managed to get the windows malware off my main drive and I can't check in a VM .. I've not gone back to check. I also haven't tried 10 bit encodes as I have a skylake CPU and that doesn't support it.

test with something like the following (intel chip and mfx libs required ofc). Note that if your library / hardware supports ICQ then CQP won't be default but CQP is by far the lower bitrate method:

HandbrakeCLI -i in.mp4 -o outcqp_byr.mkv -e qsv_h265 -q 24 -E copy -x "force-cqp=1"

Regards .. and thanks for the great software!

Dan Oram.

_Lengthy and boring notes on why QP vs ICQ for HEVC:_

For minimal bitrate it's best to use CQP with hierarchical B frames but this has high latency encode / decode so for low latency e.g. recording / editing use ICQ using library defaults.

For HEVC QP with hierarchical B layers usually produces significantly lower bit rates than HEVC ICQ or x264 from the same source with almost no visual degradation as it provides an offset of n to QP when encoding a frame in the nth layer of the B tree thus favouring bitrate for encoding information that persists for longer periods of time. The results are broadly aligned with the effects of time and motion on uncertainty as the hierarchy is a kind of upside down binary tree where matches can occur from lower to higher layers but not the other way around. This forces lower QP on frames which affect other frames for longer periods of time. Although this can create visible effects when paused, it tends to be invisible when running at a reasonable fps as every other frame is from a lower layer and will not use a match with the higher QP region. The improvement especially for high resolution such as 4k can be mind blowing (mainly if you tweak the QPOffset .. not possible with just these patches). The ICQ measure is designed to be stable between images and is not suited to this layer based offset, and enabling the B tree will make almost no difference to bitrate with that.

DOram424 committed:
LinGUI: Add the "more settings" box for qsv encodes

These encoders support parameters and this brings the linux gui in line with the windows one.

DOram424 committed:
qsv: Use appropriate aliases for presets

Use balanced, speed, best aliases for preset instead of targetusage_n as the inbetween n values (e.g. 2) are undefined and the release notes for the latest intel media sdk explicitly claim these should return errors. They don't return errors, but this is clearer and better defined.

DOram424 committed:
qsv: Add support for idr-interval param

Controls whether I frames are IDR or not. For HEVC value of 0 means only the first I frame is Idr wheras for h264 it means all frames are Idr. Makes output files non-seekable for some HEVC players and makes basically no difference.

DOram424 committed:
qsv: Increase max gop-ref-dist size to 256 for HEVC

The HEVC codec uses a hierarchical B pyramid for best results and this requires a long GOP to best target bitrate reduction at information that persists for significant period of times. The HEVC spec. allows a length of up to 256 (roughly 10 seconds at 30fps).

DOram424 committed:
qsv: Enforce GOP parameter constraints for HEVC

The hierarchical B pyramid used in HEVC requires the GOP size be a power of 2 and at most 256. Furthermore, the number of reference frames for the B pyramid is defined in the unsupported CodingOption3 so must be fixed at the max defined for HEVC of 15. If not using a b pyramid these constraints do not apply and the hw default can now be obtained by providing 0s or leaving them unset.

CQP defaults to b pyramid if using HEVC wheras ICQ defaults to it's simple frequent GPB frame detection. No constraints are applied to ICQ .. defaults are ideal.

DOram424 committed:
qsv: Detect supported params for hw codecs

All encoders take all structures, they just tend to use radically different parameters most of which get reset on the call to init if they are supplied invalid. This change updates the detection code to use this reset to detect support and allows HEVC to use the CodingOption2 structure to configure the B pyramid. This fix is minimal and is only applied to parameters needed for HEVC.

sr55 commented:
@maximd33  any comments?

maximd33 commented:
for sure @sr55 ,
allow me come back to these changes shortly

@DOram424 thanks for the contribution and while reviewing many changes: do you have results to share about : 

> sample_encode executable that is part of the mfx libraries was producing smaller files with similar quality

?

DOram424 commented:
Sorry for the oversight. Here are some sample results from re-encoding a 2 minute clip (resolution 1920x1080 ... comp. graphics swarming bees .. difficult encode) with the source recorded at very high quality. You should find all but the least stable (in time and space) of videos should encode to similar size improvements. By that constraint very low res. or very short videos may show very little improvement. Strictly speaking .. the q values are not directly comparable but they are roughly comparable and all encodes show no visible artifacts even when paused (that I can see ..).

_Hardware encode (post patch):_

HandbrakeCLI -i swarm.mp4 -o swarm_qsv_h265_q24.mkv  -e qsv_h265 --encoder-preset quality -q 24 -E copy -x "force-cqp=1"
Encode Time: 1min 35s  File Size: 79.3Mb 

_Hardware encode (pre patch):_

HandbrakeCLI -i swarm.mp4 -o swarm_qsv_h265_q24.mkv  -e qsv_h265 --encoder-preset quality -q 24 -E copy -x "force-cqp=1"
Encode Time: 1min 25s  File Size: 121.6mb

_Software encode_

HandbrakeCLI -i swarm.mp4 -o swarm_x265_q24.mkv  -e x265 --encoder-preset medium -q 24 -E copy

Encode Time: 5min 28s File Size: 86.4mb

Note that pre-patch it is possible to get better results using parameters, but full benefit cannot be achieved due to constraints on their maximum values, and there may be playback problems on some players due to IDR frames (output is perfectly valid though). A perfectly acceptable alternative to the proposed patches would be to break the limits and provide presets. That basically removes the largest sub-patch that fixes parameter detection / validity checking for HEVC.

_Hardware encode improved (pre-patch):_

HandbrakeCLI -i swarm.mp4 -o swarm_qsv_h265_q24.mkv  -e qsv_h265 --encoder-preset quality -q 24 -E copy -x "force-cqp=1:gop-ref-dist=16:gop-pic-size=256:num-ref-frame=15"

Encode Time: 1min 34s   File Size: 90.1mb

DOram424 commented:
And to reference sample_encode .. look here

https://github.com/Intel-Media-SDK/MediaSDK/tree/master/samples/sample_encode

A little new to github. Please set me right if I'm doing things wrong.

DOram424 commented:
Fixed a redundant #if in the patch ...

galinart commented on libhb/handbrake/qsv_common.h:
```diff
@@ -226,7 +226,8 @@ typedef struct QSVFrame {
 
-#define HB_POOL_SURFACE_SIZE (64)
+// Large size is driven by hevc which has a maximum GOP size of 256
+#define HB_POOL_SURFACE_SIZE (268)
```
As I remember HB_POOL_SURFACE_SIZE will be used as ArraySize parameter in D3D11_TEXTURE2D_DESC.
https://docs.microsoft.com/en-us/windows/win32/api/d3d11/ns-d3d11-d3d11_texture2d_desc

I am not sure that it is possible to create so large array in the texture, that is why two pools were created in EncQSVFramesContext.. but let me check

DOram424 replied:
Thanks .. DirectX isn't something I have much experience with .. and the pool size is fine under linux so that is a potential problem that could easily have escaped my cursory windows testing. It did however work for me under windows with a post patch HandbrakeCLI.  As the docs you link claim the max array size is 2048 I was thinking that if it's not because the problem is fixed elsewhere it might have something to do with out of internal graphics memory type errors. I think it's only intel chips with Iris graphics which have a small dedicated video memory .. for all HD graphics chips (like my skylake) it's all system memory so won't run out.

I think the use of internal graphics memory is decided here as the called function reaches into the windows specific code.

https://github.com/HandBrake/HandBrake/blob/b34a00ee0e33e2400be889cf20aa6643ead5a6be/libhb/enc_qsv.c#L1036

Limits to what I can do without installing windows. If you get back to me saying you've got a concern then I promise to install windows 10 and spend some time figuring out the frame pooling code.

maximd33 replied:
we need to come up yet with proper solution about amount of resources allocated and really needed.

ffmpeg does require additional param for the cases when bigger(>64) pool is really needed: https://github.com/FFmpeg/FFmpeg/blob/6434b84460bb8b996bc52b33b2012e6b05d3a459/fftools/ffmpeg_qsv.c#L96

galinart replied:
It does not allow to create so large d3d11 texture underneath in FFMPEG. I tested it with latest Windows driver on Intel NUC. 64 is a maximum number.

DOram424 replied:
@agalin89 Have submitted a very simple test fix for this. As I can't reproduce the errors you were getting could you please test if for me. See the comment:

https://github.com/HandBrake/HandBrake/pull/2386#issuecomment-564429835

Thanks!

galinart reviewed (commented):
@DOram424 thank you for contribution, I will check it shortly and provide more detailed feedback.

maximd33 commented on libhb/qsv_common.c:
```diff
@@ -645,6 +646,54 @@ static int query_capabilities(mfxSession session, mfxVersion version, hb_qsv_inf
+                videoParam.ExtParam[0] = (mfxExtBuffer*)&extCodingOption2Out;
+                videoParam.NumExtParam = 1;
+
+                status = MFXVideoENCODE_Query(session, &inputParam, &videoParam);
```
it seems that this call is not always successful, 

@DOram424 no "check failed" in your tests?
@agalin89 can you pls suggest changes for latest drivers while on Windows?

galinart replied:
yes, it failed with the following message:
hb_qsv_info_init: mfxExtCodingOption2 check failed (0x43564548, 0x104, -3) for lookahead rate control

LA is available in two ways:
•	As a part of AVC encoder
•	As a standalone pre-analysis plug-in

Please follow https://github.com/Intel-Media-SDK/MediaSDK/issues/1743 
Shortly to use LA for HEVC codec you need to have two sessions and load plugin with ID MFX_PLUGINID_H264LA_HW in a separate session. Please close the sessions after query calls.

DOram424 replied:
The check failed stuff should be fixed now if you'd like to try again. I can be a bit of a space cadet, so please do.

As to the lookahead stuff. I'm not 100% sure I understand your concern here. Correct me if I'm wrong/misunderstanding but currently handbrake only supports lookahead for x264 encodes and the plugin version is .. just not supported.

It's out of scope for this patch to change the behaviour of handbrake wrt to lookahead but 
I may have a future patch where it becomes in scope so am very interested to hear if you found any noticeable improvements from lookahead with hevc B pyramid encodings and long gops (128+). Given the bi-directional inter-frame matching I'd expect a lot less benefit compared to x264.

galinart commented on libhb/enc_qsv.c:
```diff
@@ -1342,54 +1413,75 @@ int encqsvInit(hb_work_object_t *w, hb_job_t *job)
+        if (!pv->is_sys_mem)
+        {
+            // limit these to avoid running out of resources (causes hang)
+            pv->param.videoParam->mfx.GopRefDist   = FFMIN(pv->param.videoParam->mfx.GopRefDist,
```
GopRefDist is distance between I- or P (or GPB) - key frames based on https://software.intel.com/en-us/node/628473
I do not see the case where we need to have so large number.

DOram424 replied:
Closing this conversation as other than the below "of interest" discussion .. the problems with >64 surfaces in d3dll textures and lookahead support are included in other conversations above. The following is mainly of interest and shouldn't be considered important reading.

The large number for GopRefDist (256) is the specification defined maximum allowed value for hevc and is only really useful for the highly specific case of hierarchical B frames in HEVC. The huge improvements for inter-frame matching for HEVC allow long tracks and 256 frames is about 10 seconds at 30fps. It isn't uncommon for video to have little movement for large parts of the screen or some object of central focus for 10 seconds (e.g. a zoom on talking people) . Because information that is visible for short periods is much less visually important than information visible for longer periods the B frame uses less bitrate on short lived information. This is quite intuitive as you can't read many letters if they are visible for <5-10 frames .. but can be understood less intuitively as the relationship between measurement, uncertainty, time, space and scale. For the handbrake case .. higher frequency information in time and space is generally (a) less visible and (b) can have a disproportionate effect on QP despite not being easily observed.

To get a feel .. experiment with powers of 2 in gop-ref-dist in the below cmd line (adjust the q value so you're starting to get some visual artifacts):

`HandBrakeCLI -i TourDeFrance.mpv  -o test.mkv  -e qsv_h265 --encoder-preset quality -q 26 -E copy -x "force-cqp=1:gop-ref-dist=32:gop-pic-size=512"  --disable-qsv-decoding
`

and you'll hopefully see  improvement in file sizes coming at little visual cost during playback (it may be more visible if paused). To get the biggest effect use simple animation inputs (e.g. some modern and all older cartoons or anime) and expect less effect in video with constant scene changes and high frequency data (higher resolution computer games recordings or some music videos spring to mind). I've even seen improvements of order 30-40x on 4k video for hierarchical QP HEVC vs ICQ HEVc.

galinart commented on libhb/handbrake/qsv_libav.h:
```diff
@@ -158,7 +158,9 @@ THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-#define HB_QSV_SURFACE_NUM              80
+// You can blame HEVC encoding as it has a maximum gop size of 256.
+#define HB_QSV_SURFACE_NUM              268
```
Main setting which defines number of frames is a number of B frames (GopRefDist), not GopSize

DOram424 committed:
qsv: Fix downsampling checks

Downsampling is only used by lookahead bitrate methods so check only if lookahead is detected as working. Also de-active debugging code that had been accidentally left on.

DOram424 commented:
Fixed some of the above problems. Apologies .. really shoddy work on my behalf there - quite a bit of testing / checking code was left on.

DOram424 committed:
Fix memory problems when using qsv decoding:

(I can't reproduce the memory errors this attempts to fix and so this is DRAFT COMMIT). Lots of tidying still needed ...

Allow decoder to work with max 64 frames. If encoder needs more than 64 frames force the encoder to take system memory frames instead of system specific. Also prevent encoder from crashing when more than 256 frames have been supplied to encoder as that is the max cached (i.e. locked) frames the encoder can handle.

DOram424 commented:
Applied a DRAFT patch to fix the frame allocation and resource constraint problems .. it is draft as I do not get these bugs pre or post patch and I'll tidy it up when it's confirmed to work.  It would be great If somebody who had problems could test this and let me know if it works post patch for HEVC encode. Example command lines below.

The patch basically seperates the decode frame cache from the encode frame cache by setting is_sys_mem=1 in the encoder if encoding is going to be high memory use.  N.B. The encoder cannot have more than 256 frames outstanding .. hence the async depth logic.

Dan.

To test: the following both with and without disable-qsv-decoding switched to enable should work. Previously the enable-qsv-decoding versions failed for me in some cases.

`HandbrakeCLI -i Dodogama.mkv -o outcqp_byr.mkv -e qsv_h265 -q 24 -E copy -x "force-cqp=1" --verbose=10  --disable-qsv-decoding`

`HandbrakeCLI -i Dodogama.mkv -o outcqp_byr.mkv -e qsv_h265 -q 24 -E copy -x "force-cqp=0" --verbose=10  --disable-qsv-decoding`

sr55 commented:
Just checking in for updates on this?
